content/en/blog/_posts/2023-03-08-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,269 @@
+
+* `spec.containers[].resources.limits`
+
+  When you specify the resource limit for containers in a Pod, the kubelet enforces 
+  those limits so that the running containers are not allowed to use more of those 
+  resources than the limits you set.
+
+When the kubelet starts a container as a part of a Pod, kubelet passes the 
+container's requests and limits for CPU and memory to the container runtime.
+The container runtime assigns both CPU request and CPU limit to a container.
+Provided the system has free CPU time, containers are guaranteed to be 
+allocated as much CPU as it requests. Containers cannot use more CPU than
+the configured limit. However, container runtimes don't terminate containers
+for excessive CPU usage, and rather throttle the CPU.
+
+Prior to Memory QoS feature, the container runtime assigns only Memory limit
+to a container, and did not use Memory request configuration. If a container
+uses more Memory than the configured limit, the Linux kernel activates Out
+Of Memory (OOM) subsystem to kill the process in the container that tried
+to allocate the extra memory. 
+
+Let's compare how the container runtime on Linux typically configures Memory
+request and limit in cgroups, with and without Memory QoS feature:
+
+* **Memory request**
+
+  The memory request is mainly used by kube-scheduler during (Kubernetes) Pod 
+  scheduling. In cgroup v1, there are no controls to specify the minimum amount
+  of memory the cgroup must always retain. Hence, the container runtime did not
+  use the value of requested memory set in the Pod manifest. 
+
+  cgroups v2 introduced a `memory.min` setting, used to specify the minimum amount of memory
+  a cgroup must always retain. If the memory usage of a cgroup is within its effective
+  min boundary, the cgroup’s memory won’t be reclaimed under any conditions. If 
+  there's no unprotected reclaimable memory available, the kernel invokes its 
+  OOM killer. Memory QoS maps `memory.min` to `spec.containers[].resources.requests.memory`
+  to ensure the availability of memory for containers in Kubernetes Pods. 
+
+* **Memory limit**
+  
+  The memory limit specifies the memory limit, beyond which if the container tries to allocate 
+  more memory, Linux kernel will terminate it with an OOM (Out of Memory) kill.
+
+  In cgroup v1, `memory.limit_in_bytes` interface is used to set the memory usage limit.
+  However, unlike CPU memory can't be compressed as cgroup v1 doesn't provide any interface
+  to specify memory usage throttle limit. 
+
+  In cgroup v2, `memory.max` is analogous to `memory.limit_in_bytes` in cgroupv1.
+  And `memory.high` interface is introduced to specify memory usage throttling
+  limit. Memory QoS uses `memory.max` to specify memory limits, and `memory.high` 
+  to throttle workload approaching its memory limit, ensuring that the system is
+  not overwhelmed by OOM kills from containers trying to use memory beyond specified
+  limits.  
```

Comment:
I would repharse this a bit. MemoryQoS cannot avoid OOM for workloads that continuously allocate memory. how about something like:

> ensuring that as containers reach their `memory.high` limits, the kernel is able to reclaim memory which may avoid an OOM.

Reply:
PTAL @bobbypage

Reply:
Thank you for update looks great!